modules/mono/glue/GodotSharp/GodotSharp/Core/GodotSynchronizationContext.cs

```diff
@@ -1,17 +1,44 @@
 using System;
 using System.Collections.Concurrent;
-using System.Collections.Generic;
 using System.Threading;
+using System.Threading.Tasks;
 
 namespace Godot
 {
     public sealed class GodotSynchronizationContext : SynchronizationContext, IDisposable
     {
-        private readonly BlockingCollection<KeyValuePair<SendOrPostCallback, object>> _queue = new();
+        private readonly BlockingCollection<(SendOrPostCallback d, object state)> _queue = new();
+
+        public override void Send(SendOrPostCallback d, object state)
+        {
+            // Shortcut if we're already on this context
+            // Also necessary to avoid a deadlock, since Send is blocking
+            if (Current == this)
+            {
+                d(state);
+                return;
+            }
```

Comment:
Doesn't this mean the callback would be invoked out of order? Since it would skip the queue. I'm not sure how important keeping the order is.

Reply:
Ah yep this was always something that had concerned me as well before, but it has been a sort of long-established pattern in UI libraries (which also implement the ability to synchronously and asynchronously queue work onto a primary application thread):

- [WinForms](https://referencesource.microsoft.com/#System.Windows.Forms/winforms/Managed/System/WinForms/Control.cs,7833) (microsoft)
- [WPF](https://referencesource.microsoft.com/#WindowsBase/Base/System/Windows/Threading/DispatcherSynchronizationContext.cs,61) (microsoft)
- [Avalonia](https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Base/Threading/AvaloniaSynchronizationContext.cs#L39) (third party)
- [Nito.AsyncEx](https://github.com/StephenCleary/AsyncEx/blob/master/src/Nito.AsyncEx.Context/AsyncContext.SynchronizationContext.cs#L56) (third party)

This had always bothered me until I realized that it wasn't really "queueing up the callback with the highest priority and skipping the line", but rather "handling the case where synchronization isn't actually needed". The main purpose of `Send` is to allow _other_ threads to queue up work on the main thread and wait for completion. The case that the caller _isn't_ actually on another thread is outside the context of the actual problem, but since it is something that is so often done by accident by callers, this has been traditionally handled with this sort of short-circuit logic. The more "correct" way to handle it in my opinion would be to throw an `InvalidOperationException` 😁, but I think it has just traditionally been decided that this would cause more headaches to the users than not, I think Godot should follow suit, as it would be unexpected if we changed such a long-standing pattern.

Reply:
That makes sense. Should we be capturing the current `SynchronizationContext` and setting it then restoring it?

```csharp
var original = Current;

// ...

_queue.Add((st =>
{
    try
    {
        SetSynchronizationContext(this);
        d(st);
    }
    finally
    {
        SetSynchronizationContext(original);
        source.SetResult();
    }
}, state));
```

Reply:
I think we're good actually, since the entire lambda in that snippet (including the try/catch) are already executing inside the main thread from `ExecutePendingContinuations`, so:
```csharp
// Shortcut if we're already on this context
// Also necessary to avoid a deadlock, since Send is blocking
if (Current == this)
{
    d(state);
    return;
}

// If we've made it here, we know that we're on a non-Godot thread

var source = new TaskCompletionSource();

_queue.Add((st =>
{
    // Inside here, we are back on the Godot thread,
    // since ExecutePendingContinuations is executing the callbacks in the _queue

    try
    {
        // The current context is already the Godot context, so no need to switch
        d(st);
    }
    finally
    {
        source.SetResult();
    }
}, state));

// This will block the non-Godot thread until source.SetResult() is called from the Godot thread
source.Task.Wait();
```

Reply:
Also side question, do you know of any plans to add unit tests to this project that run as part of the main C++ test suite? It would be very helpful to have unit tests for stuff like this, as multi-threading is often hairy and unit tests can help keep things in order

Reply:
> I think we're good actually, since the entire lambda in that snippet (including the try/catch) is already executing inside the main thread from `ExecutePendingContinuations`

Ah yes, you are right.

> do you know of any plans to add unit tests to this project that run as part of the main C++ test suite?

I'm not sure if they'll be run as part of the C++ tests, but we do want to add C# unit tests eventually. It may take a while though, I don't think there's any concrete plans for it yet.